### ISSUE TITLE
Feedback: Callback Usage

### ISSUE THREAD

**Issue by dasjestyr** [Particular/docs.particular.net#2285]
Page: /samples/callbacks/sample.md

Example where I would have used SendLocal() ?

**Comment by SimonCropp** [comment]
@dasjestyr 

SendLocal is essentially "Sending to self" https://docs.particular.net/nservicebus/messaging/send-a-message#sending-to-self

So in the context of Callback it doesnt really apply

could you perhaps elaborate a little?

**Comment by dasjestyr** [comment]
1. Request comes in through a web api endpoint, I request a reply from the handler which is just trying to create an object. 
2. If the object is invalid (bad parameters; failed to construct) then the domain model throws and the handler catches it and replies that it fails and short-circuits, otherwise it replies that it is ok and continues on with the service asynchronously
3. If the client (the web controller) gets a success, it assumes it has been queued for processing and returns 202 Accepted, otherwise it knows it has not been queued and returns 400 Bad Request

So in this scenario, I need to queue it within the service that received the request (which didn't come in via the typical command channels), which I had been using SendLoca() for successfully. But I'd like to allow the domain model to validate instead of duplicating the validation at the controller, so I thought I'd try the callbacks.

**Comment by SimonCropp** [comment]
> If the object is invalid (bad parameters; failed to construct) then the domain model throws and the handler catches it and replies that it fails and short-circuits

> If the client (the web controller) gets a success, it assumes it has been queued for processing and returns 202 Accepted, otherwise it knows it has not been queued and returns 400 Bad Request

Instead of using callbacks do as much validation as possible at the web api point. then dont have the handler catch the exception. assume that the message is valid. if it isnt then let the handler throw and the message can go to the error queue for manual intervention.

> But I'd like to allow the domain model to validate instead of duplicating the validation at the controller

This may be desired but you will cause non-trivial architectural issues in an effort to achieve this

**Comment by dasjestyr** [comment]
> This may be desired but you will cause non-trivial architectural issues in an effort to achieve this

I understand. I think I came across an example that suggested that this could and/or should be done using this duplex pattern. May I ask then, what is the intent of this? What's the use case?

**Comment by SimonCropp** [comment]
@dasjestyr does this cover it https://docs.particular.net/nservicebus/messaging/callbacks#when-to-use-callbacks  ?

**Comment by dasjestyr** [comment]
Not really, or at least it isn't very clear. While it's labeled "when to you use callbacks" it seems that it only talks about *how* to use callbacks.

**Comment by andreasohlund** [comment]
>"when to you use callbacks"

In my opinion the use case for callback are when you want to use messaging behind a syncronous api that you can't change.

Example:

Old MVC app where the controller does loads of DB work that takes up to 20 seconds

#### Problem
The app has issues under load since each request will hog a webserver thread for up to 20 seconds and this causes IIS to run out of threads, server to run out of memory etc.

#### Solution
Here we can use callbacks together with the async support in asp.net ([AsyncController](https://msdn.microsoft.com/en-us/library/ee728598(v=VS.100).aspx)) etc  to start using messaging towards a new "backend" but still keeping the controller look "syncronous".

So with this trick

https://github.com/Particular/docs.particular.net/blob/master/samples/callbacks/Callbacks_1/WebSender/HomeController.cs#L43

we no longer hold the webserver thread for 20s allowing it to serve new requests as soon as the "request" message is send, with MSMQ this is a very quick local operation, should be around 20-40 ms. 

When the response comes back NServiceBus will invoke the callback which in turn will make the asp.net callback happen and the response sent back to the user.

In short: Only use callbacks when fixing legacy webapps/wcf/etc

Does this make any sense?

**Comment by dasjestyr** [comment]
A little, but:

> we no longer hold the webserver thread for 20s allowing it to serve new requests as soon as the "request" message is send, with MSMQ this is a very quick local operation, should be around 20-40 ms.

Isn't this just queuing? Why would I need a callback for that? I feel like the point of the callback would be that so the handler can respond in some way to give info back to the web application that sent the command. While it might be a bad practice, couldn't this be demonstrated by my validation example? Or am I still not seeing the picture that you're drawing?

**Comment by andreasohlund** [comment]
Your spot on, the callback is needed to keep the webrequest waiting until the backend responds with some kind of "order acked" message. This pattern frees up the webserver thread until the response comes back.

Remeber the use case is to get messaging into existing legacy code. If you built this from scratch you would go with fire and forget instead.

Does this make sense?

**Comment by weralabaj** [comment]
@dasjestyr @andreasohlund anything more to add here https://github.com/Particular/docs.particular.net/commit/8a1e3e11d4600f6a22f001434ad7f974a9d24af1?

**Comment by andreasohlund** [comment]
Looks like a good improvement to me

**Comment by dasjestyr** [comment]
@weralabaj I think that'll do it

@andreasohlund  yes I think that makes sense, thank you. Thanks you for taking the time to help me understand!
